Guard APIs

Sergio0694 opened the pull request:
## PR Type
What kind of change does this PR introduce?
<!-- Please uncomment one ore more that apply to this PR -->

<!-- - Bugfix -->
 - Feature
<!-- - Code style update (formatting) -->
<!-- - Refactoring (no functional changes, no api changes) -->
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
<!-- - Sample app changes -->
<!-- - Other... Please describe: -->

## What is the new behavior?

This PR introduces the `Guard` class, which can be used to validate method arguments in a much more streamlined manner, which is also less verbose and less error prone.

These APIs are built with three core principles in mind:

- Being **FAST**. To achieve that, all the APIs have been designed to produce as little code as possible in the caller, and each single `Guard` API will (almost always) be inlined. Furthermore, specialized methods are generated with T4 templates to achieve the most efficient assembly code possible when working with common types (eg. primitive numeric types).
- Being **HELPFUL**. Each `Guard` API produces a detailed exception message that clearly specifies what went wrong, along with additional info (eg. current variable values), when applicable.
- Being **INTUITIVE**. To achieve this, all the `Guard` APIs have expressive and self-explanatory names that make it immediately clear what each API is supposed to do. This shifts the burden of writing checks away from the developers, letting them express conditions using natural language.

Here is a sample of how these APIs greatly reduce the complexity of code and make it easier to write and to follow, while also making it less error prone:
```csharp
// Before
if (array is null) throw new ArgumentNullException("The array must not be null");
if (array.Length >= 10) throw new ArgumentException($"The array must have less than 10 items, had a size of {array.Length}");
if (index < 0 || index >= array.Length) throw new ArgumentOutOfRangeException($"The index must be in the [0, {array.Length}) range, was {index}");

// After
Guard.IsNotNull(array, nameof(array));
Guard.HasSizeLessThan(array, 10, nameof(array));
Guard.IsInRangeFor(index, array, nameof(index));
```

Below is a breakthrough of most of the new types and APIs added in this PR (there are more):

<details>
  <summary><b>Click to expand!</b></summary>

```csharp
namespace Microsoft.Toolkit.Diagnostics
{
    public static class Guard
    {
        public static void IsNull<T>(T? value, string name) where T : class;
        public static void IsNull<T>(T? value, string name) where T : struct;
        public static void IsNotNull<T>(T? value, string name) where T : class;
        public static void IsNotNull<T>(T? value, string name) where T : struct;
        public static void IsOfType(object value, string name);
        public static void IsOfType(object value, Type type, string name);
        public static void IsAssignableToType<T>(object value, string name);
        public static void IsAssignableToType(object value, Type type, string name);
        public static void IsEqualTo<T>(T value, T target, string name) where T : notnull, IEquatable<T>;
        public static void IsNotEqualTo<T>(T value, T target, string name) where T : notnull, IEquatable<T>;
        public static void IsBitwiseEqualTo<T>(T value, T target, string name) where T : unmanaged;
        public static void IsReferenceEqualTo<T>(T value, T target, string name) where T : class;
        public static void IsReferenceNotEqualTo<T>(T value, T target, string name) where T : class;
        public static void IsTrue(bool value, string name);
        public static void IsFalse(bool value, string name);
        public static void IsLessThan<T>(T value, T max, string name) where T : notnull, IComparable<T>;
        public static void IsLessThanOrEqualTo<T>(T value, T maximum, string name) where T : notnull, IComparable<T>;
        public static void IsGreaterThan<T>(T value, T minimum, string name) where T : notnull, IComparable<T>;
        public static void IsGreaterThanOrEqualTo<T>(T value, T minimum, string name) where T : notnull, IComparable<T>;
        public static void IsInRange<T>(T value, T minimum, T maximum, string name) where T : notnull, IComparable<T>;
        public static void IsNotInRange<T>(T value, T minimum, T maximum, string name) where T : notnull, IComparable<T>;
        public static void IsBetween<T>(T value, T minimum, T maximum, string name) where T : notnull, IComparable<T>;
        public static void IsNotBetween<T>(T value, T minimum, T maximum, string name) where T : notnull, IComparable<T>;
        public static void IsBetweenOrEqualTo<T>(T value, T minimum, T maximum, string name) where T : notnull, IComparable<T>;
        public static void IsNotBetweenOrEqualTo<T>(T value, T minimum, T maximum, string name);
        public static void IsNullOrEmpty(string? text, string name);
        public static void IsNotNullOrEmpty(string? text, string name);
        public static void IsNullOrWhitespace(string? text, string name);
        public static void IsNotNullOrWhitespace(string? text, string name);
        public static void IsEmpty(string text, string name);
        public static void IsNotEmpty(string text, string name);
        public static void IsWhitespace(string text, string name);
        public static void IsNotWhitespace(string text, string name);
        public static void HasSizeEqualTo(string text, int size, string name);
        public static void HasSizeNotEqualTo(string text, int size, string name);
        public static void HasSizeAtLeast(string text, int size, string name);
        public static void HasSizeAtLeastOrEqualTo(string text, int size, string name);
        public static void HasSizeLessThan(string text, int size, string name);
        public static void HasSizeLessThanOrEqualTo(string text, int size, string name);
        public static void IsEmpty<T>(T[] array, string name);
        public static void IsNotEmpty<T>(T[] array, string name);
        public static void HasSizeEqualTo<T>(T[] array, int size, string name);
        public static void HasSizeNotEqualTo<T>(T[] array, int size, string name);
        public static void HasSizeAtLeast<T>(T[] array, int size, string name);
        public static void HasSizeAtLeastOrEqualTo<T>(T[] array, int size, string name);
        public static void HasSizeLessThan<T>(T[] array, int size, string name);
        public static void HasSizeLessThanOrEqualTo<T>(T[] array, int size, string name);
        public static void HasSizeEqualTo<T>(T[] source, T[] destination, string name);
        public static void HasSizeLessThanOrEqualTo<T>(T[] source, T[] destination, string name);
        public static void IsEmpty<T>(IEnumerable<T> enumerable, string name);
        public static void IsNotEmpty<T>(IEnumerable<T> enumerable, string name);
        public static void HasSizeEqualTo<T>(IEnumerable<T> enumerable, int size, string name);
        public static void HasSizeNotEqualTo<T>(IEnumerable<T> enumerable, int size, string name);
        public static void HasSizeAtLeast<T>(IEnumerable<T> enumerable, int size, string name);
        public static void HasSizeAtLeastOrEqualTo<T>(IEnumerable<T> enumerable, int size, string name);
        public static void HasSizeLessThan<T>(IEnumerable<T> enumerable, int size, string name);
        public static void HasSizeLessThanOrEqualTo<T>(IEnumerable<T> enumerable, int size, string name);
        public static void IsEmpty<T>(ReadOnlySpan<T> span, string name);
        public static void IsNotEmpty<T>(ReadOnlySpan<T> span, string name);
        public static void HasSizeEqualTo<T>(ReadOnlySpan<T> span, int size, string name);
        public static void HasSizeNotEqualTo<T>(ReadOnlySpan<T> span, int size, string name);
        public static void HasSizeAtLeast<T>(ReadOnlySpan<T> span, int size, string name);
        public static void HasSizeAtLeastOrEqualTo<T>(ReadOnlySpan<T> span, int size, string name);
        public static void HasSizeLessThan<T>(ReadOnlySpan<T> span, int size, string name);
        public static void HasSizeLessThanOrEqualTo<T>(ReadOnlySpan<T> span, int size, string name);
        public static void HasSizeEqualTo<T>(ReadOnlySpan<T> source, Span<T> destination, string name);
        public static void HasSizeLessThanOrEqualTo<T>(ReadOnlySpan<T> source, Span<T> destination, string name);
        public static void IsEmpty<T>(ReadOnlyMemory<T> memory, string name);
        public static void IsNotEmpty<T>(ReadOnlyMemory<T> memory, string name);
        public static void HasSizeEqualTo<T>(ReadOnlyMemory<T> memory, int size, string name);
        public static void HasSizeNotEqualTo<T>(ReadOnlyMemory<T> memory, int size, string name);
        public static void HasSizeAtLeast<T>(ReadOnlyMemory<T> memory, int size, string name);
        public static void HasSizeAtLeastOrEqualTo<T>(ReadOnlyMemory<T> memory, int size, string name);
        public static void HasSizeLessThan<T>(ReadOnlyMemory<T> memory, int size, string name);
        public static void HasSizeLessThanOrEqualTo<T>(ReadOnlyMemory<T> memory, int size, string name);
        public static void HasSizeEqualTo<T>(ReadOnlyMemory<T> source, Memory<T> destination, string name);
        public static void HasSizeLessThanOrEqualTo<T>(ReadOnlyMemory<T> source, Memory<T> destination, string name);
        public static void CanRead(Stream stream, string name);
        public static void CanWrite(Stream stream, string name);
        public static void CanSeek(Stream stream, string name);
        public static void IsAtStartPosition(Stream stream, string name);
        public static void IsCompleted(Task task, string name);
        public static void IsCompletedSuccessfully(Task task, string name);
        public static void IsFaulted(Task task, string name);
        public static void IsCanceled(Task task, string name);
        public static void HasStatus(Task task, TaskStatus status, string name);
    }
}
```
</details>

There are also a couple of new extension methods being introduced in this PR, which are used internally. Since they were pretty general and easily reusable, they have been marked as public:

```csharp
namespace Microsoft.Toolkit.Extensions
{
    public static string ToHexString<T>(this T value) where T : unmanaged;
    public static string ToTypeString(this Type type);
}
```

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tested code with current [supported SDKs](../readme.md#supported)
- [X] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: [\#302](https://github.com/MicrosoftDocs/WindowsCommunityToolkitDocs/pull/302)
- [ ] ~Sample in sample app has been added / updated (for bug fixes / features)~
    - [ ] ~Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)~
- [X] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [X] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [X] Contains **NO** breaking changes


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected -->


## Other information

This PR is a follow up from a conversation with @michael-hawker over on the [UWP Community Discord server](https://discord.gg/zBA5aCn). The `DebugClass` is missing for now, but since it should be exactly identical to the `Guard` class, just with the added `[Conditional("DEBUG")]` attribute, I was thinking we could probably use a T4 template (or some similar technique) to just auto-generate this class by copying the `Guard` code and adding that single attribute to the new class. I'm not personally familiar with setting up a T4 template, so I would need some help to get that figured out. I'll work on the `Guard` class in the meantime 👍

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Guard.String.cs:
```diff
@@ -0,0 +1,233 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void IsEmpty(string text, string name)
+        {
+            if (text.Length != 0)
```
Shouldn't we also be worried about null here?

Sergio0694 replied:
That's not needed, as the entire class is using nullable reference types 😄

All the methods that accept `null` parameters are marked with the nullable attribute, and the others leave that check to the user. This is mostly done for clarity and for performance reasons, as otherwise we'd have to add redundant null checks basically everywhere.

Eg. if you're working on a `string` that could be `null`, you'd just need to call `Guard.IsNullOrEmpty(string, string)` instead of just `Guard.IsEmpty(string, string).`

Besides, if you passed a `null` value there you'd still end up with the same `NullReferenceException` anyway, so the actual code execution would stay the same.

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Guard.String.cs:
```diff
@@ -0,0 +1,233 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void IsNotEmpty(string text, string name)
+        {
+            if (text.Length == 0)
```
Shouldn't we also be worried about null here?

Sergio0694 replied:
Same as comment above ([link](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131#discussion_r390925589)) 👍

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Guard.Comparable.Generic.cs:
```diff
@@ -0,0 +1,371 @@
+             * messages, since the entire input values can be expressed as hexadecimal values.
+             * The conditional branches below are known at compile time by the JIT compiler,
+             * so that only the right one will actually be translated into native code. */
+            if (typeof(T) == typeof(byte) ||
```
Since we're using the T4 stuff already, would it not be faster to have method overloads for the compiler to pick the right path combination vs. having ifs at runtime?

Sergio0694 replied:
It doesn't actually make a difference in this case, as those checks are only available in IL. Once the JIT compiles the method (or .NET Native on UWP), you'd get a different type-specialized method for each different type, just like if we had used a T4 template in this ase (yay reified generic types! 😄).

You can double check that the final asm is exactly the same, I've setup a sample file for you [here](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8B6YgAgBkBLMGAO1xgBMyMJWALGMgOgDkAogBUyAMQgBXOk2wYqEOmSlMYUMgu7Qy+aN2wBzWDHz0MuHgFgAUKTJCuvQSPHLZ8xQBsa9Rrk5U/ADMqD242MgBPSSVpVU5uAFkASREvWgYYSxtyAGUYMIcKJIBhAT4cgTJg0LIqRQwHAAcoCAArGDAMMhaILsCtHVhaun6ofDcFLOsAAQAmAEYZgAYyafmeACUpORMeYoh8RpDVPKgAN29cAG5lshyObChGimxgTe2qExurGdmAZmsAG9rGRQas/qt5khVig7DBcBh5gAKHKNbB0AA8dQwAD4yLg0XQ0MMusAAJQgsHAqxg2lVXTYMAcMhI7FkbDDfGEik0ul8gDiEkeTB4SVwACEqBgAO4BGACACOQo8QggyOwxOAxLo2BMEECSOwZLJ31pAF9KaDLeDIdDprChPCMLMUYSsXRcVz0cS2eTrdS+aDRjBGczWR72ZyCeieYG6YLhaKJVLZYxFcrVShDZrtbqYPrDcbTWCLVZSzMIWtoWioHJsB5VrMyAmoEwgWDrQBtABSUv59FUNCRGAijXzBuxxoAul2EjAGhAmElDh4kXOF0uVwB5RpyBTmACC+kM8NwVDOMCSdC8dDq+jJM95oOmlahMLIYslMrl6frqvmGJCDiSIiGc9YSDAxIiBgjz6POxJrCsOomLGgbSlwQwiCAMTjDqcFtk+ZABoGVCBCyI5jgWQhkmQAC8tGsKO47hhgxrWrSxFxqCEhsmBHgQQAqleXQMQJDDYIEmQHrggHEjxHrAbAZF8RBJrsYG8ldDBUBwRgQkRqJ4mSTw0myUo2KKTAZHabpak/IRcakSyKkwPpXQAIQMTZ85uahXFEep/n2C00oHjpEgmB6AgILQu7uOIUDJt+aZKn+EBIi5xLeRguYocW/mlnGhWBjAHiMIFAUOYGSlkMAEQYNwLkbFZdHdC1Ym4BJUkyUImr1TAlnKeBMB2f5oI1XVDWsLB87NWRDE1R1XUmT1fUNYN006fOo1jb6/W4Ps2ytUtxk5FQABeMBboEgHAXZFW0qMLJslQrVLFctRkBitX7YdHofVQADUgN+VxnFjaCk2NcN4r9cdRlSUwTBIjVTVWT6O0Q5DcPZbDU2GZ1xkHkjKMtdlc0Y98D0kWRGUw3DnmbbpeMjdTfLg1jYLBRAoXhZFGDRbFe50AlSWpvKqUqulmVM/BZDISN+VY8VBUVSrZDq7O84cIuy6NKu6465u+s7sLh7HrAuBnheV43neD7Wi+trvp+KY/pLmYsWQMtstlCHzEheagxzYJOXT/HcIz2WgxxbN2BwIVhfoEWmILMBxQootfuLv5S+HEFZTNOXy0HSvmtaZq0taWsbnrBva7r24Zww/AQLbdT24+tJO1W77c7zyf82nzdZ27KUZhAt0gd7w1QbLxeISXKEVehqjcFhOHogYzD+hVDQhfLMDSmQScp1FMXp8LSIACQAEQAAqPHmDVqICCuV0igIUcx1FmjRATy16JGYAY9uDjAwEyYk0o8BfUBC5M0eI6gIhDCwfUsDsoINvrlRWFdrCliAA). You can see that `Test1`, using the generic method, and `Test2` (type-specific) produce the same final native code.

Also keep in mind that for types that can be bitwise-compared directly (eg. integer types), you can also use `Guard.IsEqualTo`, that'll work as well. The main difference is just in the error message, as `IsEqualTo` will display values in base 10, whereas `IsBitwiseEqualTo` will display them in hex format.

Another reason why I went with a single generic here was to try not to have another file with 1000 lines of code just for this single API 😅

michael-hawker replied:
Yay for smart compilers! :)

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Generated/TypeInfo.ttinclude:
```diff
@@ -0,0 +1,96 @@
+    /// Gets the list of available enumerable types to generate APIs for
+    /// </summary>
+    static readonly IReadOnlyList<EnumerableTypeInfo> EnumerableTypes = new[]
+    {
```
Isn't `IEnumerable` missing here?

Sergio0694 replied:
I had it too initially but then I removed it because I didn't think it made sense. All the `Guard` APIs mostly use collections to check their size, and that's not possible with `IEnumerable<T>`. You'd need to create the enumerator and traverse the sequence, but that could get really expensive if the enumerator is doing some non-trivial work, so it's generally bad practice. By not having them we make the intent clearer and force users to first get an actual collection to work on, so that they skip multiple enumerations. I feel like that's one of the easiest mistakes to make especially for beginners, so I thought it would be a good idea to give them a hint by not supporting the type at all here.

michael-hawker replied:
Cool, just wanted to double-check it wasn't overlooked :)

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Generated/Guard.Comparable.Numeric.tt:
```diff
@@ -0,0 +1,237 @@
+        }
+
+        /// <summary>
+        /// Asserts that the input value must not be in a given range.
```
Ah, I see the subtle differences here between InRange and InBetween... wondering how to make this easier to process, is there an existing .NET equivalent to point me to for a reference?

Sergio0694 replied:
Mh I wouldn't know of an existing .NET API that's equivalent to that I'm afraid 🤔

I guess maybe we could mention the mathematical range in the remarks for each API?

- `IsInRange` ---> **[min, max)**
- `IsBetween` ---> **(min, max)**
- `IsBetweenOrEqualTo` ---> **[min, max]**

michael-hawker replied:
Yeah, I know not everyone may be familiar with that notation, but at a glance, especially in the breakdown overview doc it'd make it quickly easy to pick-up the differences.

Sergio0694 replied:
Done in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/67fcd478ab4787eb4d57e67019059476b7d2e40b and https://github.com/MicrosoftDocs/WindowsCommunityToolkitDocs/pull/302/commits/61452b3c9d5538c6683f11aa9cf66b7de32c5d9f 👍

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Generated/Guard.Collection.tt:
```diff
@@ -0,0 +1,259 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void IsInRangeFor<T>(int index, <#=item.Type#> <#=item.Name#>, string name)
+        {
+            if ((uint)index >= (uint)<#=item.Name#>.<#=item.Size#>)
```
Need lower-bound check too, eh?

Sergio0694 replied:
Ah, it's a trick! 😄

Actually we don't, because signed integers are represented in [2-complement](https://en.wikipedia.org/wiki/Two%27s_complement) (most significant bit indicates the sign, the others represent the absolute value with inverted bits + 1, with overflow). Checking whether an index is not in a _positive_ `int` range is equivalent to:

```csharp
if (index < 0 || index >= max) throw ...
```

So, if we cast both values to `uint`, we'll get the following:

- If the index was positive, it doesn't change. Then the `>=` comparison is valid, as we want out index to only be _lower_ than the maximum possible index.
- If the index was negative, casting it to an _unsigned_ value makes it so that the most significant bit (the sign) becomes part of the absolute value. Hence, if the value was negative, it'll become `index + int.MaxValue`, which is guaranteed to be always higher than the maximum index, as that number is always positive when read in 2-complement. The same comparison still works, as we want our `uint` cast to be strictly _lower_ than the maximum.

So basically we get two checks for the price of one! 😊
This is the same trick used throughout CoreCLR as well to perform range checks.

The practical effect is that this completely removes the conditional branch in asm 🎉
Here's the version with the two conditions (RyuJit, x64):

```asm
; bool IsInRangeTwo(int i, int max) => i >= 0 && i < max;

C.IsInRangeTwo(Int32, Int32)
    L0000: test ecx, ecx
    L0002: jl L000d
    L0004: cmp ecx, edx
    L0006: setl al
    L0009: movzx eax, al
    L000c: ret
    L000d: xor eax, eax
    L000f: ret

; IsInRangeOne(int i, int max) => (uint)i < (uint)max;

C.IsInRangeOne(Int32, Int32)
    L0000: cmp ecx, edx
    L0002: setb al
    L0005: movzx eax, al
    L0008: ret
```

Shorter, and with no branching! 🚀

michael-hawker replied:
![image](https://user-images.githubusercontent.com/24302614/76455788-507e0580-6393-11ea-8878-0233c2b248d5.png)

This is what I rely on perf experts like you for 😋. Maybe just add a comment to call out that we're using an optimized trick here for future me (and others)?

Sergio0694 replied:
Absolutely! Done in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/f17aacd5a1e0f58df35536fb5d15bf9f4dcb0fcd 👍

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Generated/Guard.Collection.tt:
```diff
@@ -0,0 +1,259 @@
+<#
+});
+#>
+    }
```
Are we missing `IsNotInRangeFor` since we have so many parallels?

Sergio0694 replied:
Right, I guess it could be useful as well in some cases 😄

Done in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/818c1eb524fa3fa98993c65d1a5c572c23b7556f.

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Generated/Guard.Collection.tt:
```diff
@@ -0,0 +1,259 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void HasSizeEqualTo<T>(<#=item.Type#> <#=item.Name#>, int size, string name)
+        {
+            if (<#=item.Name#>.<#=item.Size#> != size)
```
Are we worried about null case for all these?

Sergio0694 replied:
Same as comment above ([link](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131#discussion_r390925589)), the same concept applies there. All the collections are not marked as nullable, to the IDE will show a warning for you. Plus this way we get the `Guard` APIs to only perform the same exact checks they declare in the name. If you call an API called `IsNotNull`, it will check for `null`. If you call an API called `HasSizeEqualTo`, it will do just that. This makes the whole thing as straightforward as possible as well, other than also reducing the overhead 😊

Sergio0694 commented on Microsoft.Toolkit/Microsoft.Toolkit.csproj:
```diff
@@ -14,4 +16,54 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <Service Include="{508349b6-6b84-4df5-91f0-309beebad82d}" />
```
Yup, good idea!

Done in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/5f25e180b13dc8c939a75ef828d3ddec7905a635.

Sergio0694 commented on Microsoft.Toolkit/Diagnostics/Generated/Guard.Collection.cs:
```diff
@@ -0,0 +1,1502 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+/* ========================
```
I checked them in for now because apparently VS only rebuilds them if you open one and save it, and not when you just build the solution. If there's an option for that then yeah it'd be possible not to check them in. Since I didn't find that though I've left them in so that people cloning the repo don't end up with either build errors or missing APIs because of that.
If it's possible to set that, I'm all ears! 😄

As for the outputs being in separate folders, I don't think that's necessary, since VS has a very nice display view for them, just like with `.xaml` and `.xaml.cs` files, so it bundles them together:

![image](https://user-images.githubusercontent.com/10199417/76370780-40850800-6338-11ea-9e54-2b57da0de2e4.png)

michael-hawker replied:
Yeah, I'm not sure if there's a build action or not., good to know about the VS display, it's been a bit since I've worked with T4 templates too :)

Sergio0694 commented on Microsoft.Toolkit/Diagnostics/Generated/Guard.Collection.tt:
```diff
@@ -0,0 +1,259 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Sounds good! I agree that it might not be immediately clear for everyone what a T4 template even is, or why is all that code being generated, yeah. I'll put together a draft detailing the main points of how the API is structured and I'll put in that `\Generated` folder 👍

Sergio0694 replied:
Done in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/caad2f7792c29f83163212b19045a3d3d43d05af 😊

Sergio0694 commented on Microsoft.Toolkit/Extensions/ValueTypeExtensions.cs:
```diff
@@ -0,0 +1,62 @@
+        };
+
+        /// <summary>
+        /// Returns a hexadecimal <see cref="string"/> representation of a given <typeparamref name="T"/> value, left-padded and ordered as big-endian.
```
Yeah good idea! 👍

Done in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/cf5785c71667b2389e907d22d2ef16db6fa517c2.

Sergio0694 commented on Microsoft.Toolkit/Microsoft.Toolkit.csproj:
```diff
@@ -2,6 +2,8 @@
 
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
+    <LangVersion>8.0</LangVersion>
```
No, this setting only affects code inside the package, in particular it unlocks the nullable reference types. But everyone using a lower version of C# will still be able to use the package, they just won't benefit from those additional attributes. The entire assembly is still targeting just .NET Standard 2.0, I'm not using any C# 8 exclusive types (eg. `Range`) 👍

Sergio0694 commented on UnitTests/Diagnostics/Test_Guard.Array.cs:
```diff
@@ -0,0 +1,182 @@
+        public void Test_Guard_HasSizeLessThanOrEqualToArray_ArrayFail()
+        {
+            Guard.HasSizeLessThanOrEqualTo(new int[8], new int[2], nameof(Test_Guard_HasSizeLessThanOrEqualToArray_ArrayFail));
+        }
```
Yup, those are two separate functions:

- `Guard.IsInRangeFor` checks that the input index is valid for the target collection.
- `Guard.HasSizeLessThanOrEqualTo` checks that the size of the source collection is less than or equal to that of the destination collection. Basically, this can be used to check whether you can copy items from one collection to the other.

Sergio0694 commented on UnitTests/Diagnostics/Test_Guard.cs:
```diff
@@ -0,0 +1,526 @@
+            Guard.IsInRange(1, 0, 4, nameof(Test_Guard_IsInRange_Ok));
+            Guard.IsInRange(0, 0, 2, nameof(Test_Guard_IsInRange_Ok));
+            Guard.IsInRange(3.14f, 0, 10, nameof(Test_Guard_IsInRange_Ok));
+            Guard.IsInRange(1, 0, 3.14f, nameof(Test_Guard_IsInRange_Ok));
```
Sure thing! 😄

Done in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/b840201d17edd452314eab49bb3b37e9b96fce36.

Sergio0694 commented on UnitTests/Diagnostics/Test_Guard.cs:
```diff
@@ -0,0 +1,526 @@
+        [TestMethod]
+        public void Test_Guard_IsBetween_Ok()
+        {
+            Guard.IsBetween(1, 0, 4, nameof(Test_Guard_IsBetween_Ok));
```
Replying to the range/between difference in the other comment below 👍

Regarding `IsCloseTo`, I can add those two for floating point types, good idea!

Sergio0694 commented on UnitTests/Extensions/Test_ValueTypeExtensions.cs:
```diff
@@ -0,0 +1,26 @@
+        [TestMethod]
+        public void Test_ValueTypeExtensions_ToHexString()
+        {
+            Assert.AreEqual(((byte)0).ToHexString(), "0x00");
```
Yup, the one for `Color` is this one:

https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/master/Microsoft.Toolkit.Uwp/Helpers/ColorHelper.cs#L124-L132

I've used the `To____String` naming scheme as it's the same one used in the main `Microsoft.Toolkit` package (there's also [`ToArrayString`](https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/master/Microsoft.Toolkit/Extensions/ArrayExtensions.cs#L111-L120)), so I picked `ToHexString` for consistency within this package. In fact it's the same I did for the `Type` extension as well, since that's `ToTypeString`. Personally I like the `ToHexString` more, as I think it makes the intent clearer especially for less experienced developers. As in, it's returning a hexadecimal `string`, it's not "converting" the input value to hexadecimal somehow or making some kind of other transformation. Having the naming be closer to the classic `ToString` makes it immediate to understand: "ah, it's just like `ToString`, just with a different formatting".

What do you think? 😄

michael-hawker replied:
Sounds good, maybe we can add a comment to the #3062 that we should rename the function in ColorHelper for consistency in the 7.0 release?

Sergio0694 replied:
Good idea! Done: https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3062#issuecomment-597874095 👍

Sergio0694 commented on Microsoft.Toolkit/Extensions/TypeExtensions.cs:
```diff
@@ -0,0 +1,117 @@
+    /// <summary>
+    /// Helpers for working with types.
+    /// </summary>
+    public static class TypeExtensions
```
Yes exactly, plus it also includes a couple other cool things:
- Primitive types with a C# alias (eg. `int`, `string`, etc.) will be displayed like that instead of the fully qualified name (eg. `System.Int32`)
- Value tuples will be displayed as `(T1, T2, ...)` (instead of `System.ValueTuple'2[T1, T2]`)
- Nullable value types will be displayed as `T?` (instead of `System.Nullable'1[T]`)

Basically, this method should display types just like users type them in code 😄
(plus the namespace info for non built-in types)

Sergio0694 commented on Microsoft.Toolkit/Microsoft.Toolkit.csproj:
```diff
@@ -14,4 +16,54 @@
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="System.Memory" Version="4.5.3" />
```
It's for a couple reasons:

- The `Guard` APIs has a number of overloads that support `Span<T>`, `ReadOnlySpan<T>`, `Memory<T>` and `ReadOnlyMemory<T>`, yeah.
- The package also includes the `System.Runtime.CompilerServices.Unsafe` class, which is used by a number of APIs I've added (in particular, `Guard.IsBitwiseEqualTo<T>` and `ToHexString<T>`).

All the new `Span<T>`-related APIs are also available by default both in .NET Standard 2.1 and .NET Core 3.0, so eg. if we were to eventually multitarget the `Microsoft.Toolkit` package, we wouldn't even need to reference `System.Memory` directly (just `System.Runtime.CompilerServices.Unsafe` at that point).

In general, I'd say `System.Memory` is absolutely a must have package for everyone: using `Span<T>`-related APIs should **always** be preferred over creating/copying new arrays, as the speed and memory improvements are just incredible in most cases.

In fact, after this PR is merged I'd like to change a couple of extensions from the `ArrayExtensions` class to return `Memory<T>` instead of `T[]`, as it'd make them allocation-free and much more efficient. And we should definitely also consider adding overloads to those APIs (eg. `ToArrayString`) to support `Span<T>` as well, it'd make them much more flexible.

Of course I'd be happy to help with that as well 😊

michael-hawker replied:
Cool, FYI if we change the signatures to the ArrayExtensions, we'll need to do that work for the 7.0 release as it'd be a breaking change, eh?

Sergio0694 replied:
Yeah absolutely, that's one of the reasons why I didn't touch those APIs in this PR 😊

vgromfeld replied:
@michael-hawker, I don't know if there is a tool we can use to validate the impact. 
We may be able to use the size of the sample app as a reference value (even if not 100% accurate) but I have no idea on how to do that 🙁.

michael-hawker replied:
Thanks @vgromfeld, no worries, never hurts to ask. :)

Sergio0694 replied:
I realize it's probably the best comparison, but just as a test I tried to build a blank UWP app in 3 configurations:

1. Just the app with no references. I just added a test method that simply creates a random array, loops over it and prints the values, just to have some conditions. I used explicit `throw` statements.
2. I referenced `Microsoft.Toolkit` **without** the `System.Memory` package, and replaced those `throw` expressions with `Guard` calls.
3. I referenced the current `Microsoft.Toolkit` package as in this PR.

![image](https://user-images.githubusercontent.com/10199417/76576502-8a99f500-64c2-11ea-82cf-b4b79aab2267.png)

Then I created a package for each version in Release x64 (so with .NET Native).
You can see that the difference in the final package is just about **20KB** with/without the `System.Memory` package. Which makes sense, the package itself is fairly small and only contains a few types and classes, not much else.

Hope this helps! 😊

michael-hawker replied:
Thanks @Sergio0694 and to clarify was #2 with the Microsoft.Toolkit with this PR or without the PR? (Curious about the PRs impact on the bottom-line anyway as we're having this discussion.)

Sergio0694 replied:
\#2 was with the same exact `Microsoft.Toolkit` package as this PR, just after removing the `System.Memory` reference and all the APIs that were using types from that package (eg. all the `Guard` APIs working on `Span<T>`). I did this to highlight the difference of just the `System.Memory` package, as opposed to also including the binary size increase just caused by the new `Guard` APIs and the other extensions.

Sergio0694 replied:
As per @michael-hawker's request, here's a text breakdown of the binary size in my test application (UWP, Release x64, SDK 18363):

1) **No packages:**  `24.604.285` byte
2) **`Guard` API, direct reference:** `24.710.254` byte
3) **`Guard` API, direct, no `System.Memory`:** `24.690.486` byte
4) **`Microsoft.Toolkit` from NuGet:** `24.620.377` byte

Looks like the final binary size in a UWP application is actually less when the package is referenced from NuGet as opposed to when the project is referenced directly from the Visual Studio solution. And in 4) the final size was actually smaller than both the versions of this PR, with and without the `System.Memory` package being referenced.

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Generated/Guard.Collection.tt:
```diff
@@ -1,4 +1,5 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
+﻿<#@ output extension=".g.cs"#>
```
👍 awesome, like the .g.cs extension at least for check-in if we can't solve the generation on build as at least it makes it very clear it's a generated file :)

Sergio0694 replied:
Ah, glad you liked the idea! 😄

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Generated/Guard.md:
```diff
@@ -0,0 +1,15 @@
+# T4 templates and generated APIs
+
+This folder contains a number of template files (with the `.tt` or `.ttinclude` extensions) and the generated `.cs` files generated by those templates. The template files use the T4 format, which is natively supported by Visual Studio (more info is available [here](https://docs.microsoft.com/en-us/visualstudio/modeling/code-generation-and-t4-text-templates?view=vs-2019)).
```
```suggestion
This folder contains a number of template files (with the `.tt` or `.ttinclude` extensions) and the generated `.g.cs` files generated by those templates. The template files use the T4 format, which is natively supported by Visual Studio (more info is available [here](https://docs.microsoft.com/en-us/visualstudio/modeling/code-generation-and-t4-text-templates?view=vs-2019)).
```
Quick fix for the new extension

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Generated/Guard.md:
```diff
@@ -0,0 +1,15 @@
+ - Using type-specific overloads instead of generic methods can result in faster code. For instance, T4 templates are used to generate overloads for comparison APIs (eg. `Guard.IsGreaterThan(int, int, string)`). This results in more compact and optimized code as opposed to a generic method using `where T : IComparable<T>` as type constraint.
+ - In some cases, using generic methods just isn't possible. For instance, types like `Span<T>` and `ReadOnlySpan<T>` can't be used as generic type parameters, and even if that had been possible, they don't implement an interface we could use in the generic type constraint. Using T4 templates solves this issue, as we can just have specialized method for each supported type or collection type.
+
+ ## How to make changes
```
```suggestion
## How to make changes
```
extra space

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Generated/Guard.md:
```diff
@@ -0,0 +1,15 @@
+
+ ## How to make changes
+
+If you need to change an API that is declared in a template, or to add a new one, just edit the right `.tt` file and save it: Visual Studio will take care of updating the generated `.cs` file automatically. Don't make changes to those generated `.cs` files directly, as those will be overwritten as soon as their source template is updated.
```
```suggestion
If you need to change an API that is declared in a template, or to add a new one, just edit the right `.tt` file and save it: Visual Studio will take care of updating the generated `.g.cs` file automatically. Don't make changes to those generated `.g.cs` files directly, as those will be overwritten as soon as their source template is updated.
```

Sergio0694 replied:
Done in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/ac44dadfc511deea7ffbbfa5dc4cbaabcbc7e1b3.

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Guard.Comparable.Numeric.cs:
```diff
@@ -0,0 +1,127 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void IsCloseTo(int value, int target, int delta, string name)
+        {
+            /* Cast to long before calculating the difference to avoid overflows
```
👍 Yay for amazing comment about optimization usage! :)

michael-hawker commented on Microsoft.Toolkit/Diagnostics/Guard.Comparable.Numeric.cs:
```diff
@@ -0,0 +1,127 @@
+        /// <param name="name">The name of the input parameter being tested.</param>
+        /// <exception cref="ArgumentException">Thrown if (<paramref name="value"/> - <paramref name="target"/>) > <paramref name="delta"/>.</exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void IsCloseTo(int value, int target, int delta, string name)
```
Do we need one for `long` as well?

Sergio0694 replied:
Sure thing! 😄

Done in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/41fea84172629f566aa090034ce3036ec6151312 and https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/132d240ded5d271e1e92a28c5647fefa0800b06d, and also tweaked the surface area a bit 👍

Sergio0694 commented:
Took the liberty of marking individual reviews as resolved after having performed the requested changes or provided more info on specific implementation details, just to make it easier to navigate the PR and to properly track all the various notes.

Feel free to reopen them or to leave more notes if you feel anything is missing or if you have more things you'd like me to look into! 😊

michael-hawker commented:
Thanks @Sergio0694 for all the info and changes. I'm pretty good on this PR, the only pieces I have are my open questions to my colleagues: 

1) To @azchohfi if he has any knowledge of T4 and suggestions on the build process as per my comment [here](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131#discussion_r390665471)
2) To @vgromfeld if he has any thoughts/concerns on the package structure [here](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131#discussion_r390672177)

If there's no further input or concerns from them on those topics, then I'm good to merge.

azchohfi commented on Microsoft.Toolkit/Diagnostics/Generated/Guard.md:
```diff
@@ -0,0 +1,15 @@
+
+## How to make changes
+
+If you need to change an API that is declared in a template, or to add a new one, just edit the right `.tt` file and save it: Visual Studio will take care of updating the generated `.g.cs` file automatically. Don't make changes to those generated `.g.cs` files directly, as those will be overwritten as soon as their source template is updated.
```
Since we **are** checking in the .g.cs files, I would just add a note that they need to be checked in if anyone changes the .tt files.

Sergio0694 replied:
I've added the note you mentioned in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/949134be6382f35e1e1402a8b83017336ae42dc9. Is this what you had in mind?

azchohfi replied:
Yup! Thanks!

Sergio0694 replied:
No problem, thank you for your time! 😊

michael-hawker commented:
Note to self: Squash this commit as per @azchohfi's recommendation.

@Sergio0694 thanks for this amazing addition! I'm just going to do a quick final play locally with it in the morning and then will merge the PR!

michael-hawker commented:
@Sergio0694 seeing 4 test cases fail?

IsCloseToFloat_Fail
IsCloseToInt_Fail
IsNull_ClassFail
IsNull_StructFail

Sergio0694 commented:
@michael-hawker Good catch! Fixed in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/51b45e2dd7ff8687c0d2d5ed457628d6251e090a and https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3131/commits/244da990ee3458d91bcaa8013bb064b529e85d39.

- `IsCloseToFloat_Fail` was failing due to a rounding error, `float` can't actually tell the difference between `float.MaxValue` and `float.MaxValue - 10`. Changed that to `float.MaxValue / 2`.
- `IsCloseToInt_Fail` was failing because I had accidentally used `int` as the delta, which can't be implicitly cast to `uint`, so the compiler was picking up the `IsCloseTo` overload with `float` parameters instead. And those were again failing due to insufficient precision. Fixed that by changing the delta type to `uint` in the tests.
- `IsNull_ClassFail` and `IsNull_StructFail` were indeed failing because I was watching for `ArgumentNullException` in the tests, as opposed to just `ArgumentException`.

All are passing now 😄

Sergio0694 commented:
Wo-hoo! 😄 🚀

Thank you so much @michael-hawker - so happy to have my first feature PR merged in!
Can't wait for other developers to give the new APIs a try once the new update goes live! 🎉
